src/Middleware/InitializeTenancyByRequestData.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Stancl\Tenancy\Middleware;
+
+use Closure;
+use Illuminate\Http\Request;
+use Stancl\Tenancy\Exceptions\TenantCouldNotBeIdentifiedException;
+
+class InitializeTenancyByRequestData
+{
+    /** @var callable */
+    protected $onFail;
+
+    public function __construct(callable $onFail = null)
+    {
+        $this->onFail = $onFail ?? function ($e) {
+            throw $e;
+        };
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        if ($request->method() !== 'OPTIONS') {
+            try {
+                $this->parseTenant($request);
+            } catch (TenantCouldNotBeIdentifiedException $e) {
+                ($this->onFail)($e);
+            }
+        }
+
+        return $next($request);
+    }
+
+    protected function parseTenant(Request $request)
+    {
+        if (tenancy()->initialized) {
+            return;
+        }
+
+        $header = config('tenancy.identification.header');
+        $query = config('tenancy.identification.query_parameter');
+
+        $tenant = null;
+        if ($request->hasHeader($header)) {
+            $tenant = $request->header($header);
+        } elseif ($request->has($query)) {
+            $tenant = $request->get($query);
+        } elseif (! in_array($request->getHost(), config('tenancy.exempt_domains', []), true)) {
+            $tenant = explode('.', $request->getHost())[0];
```

Comment:
I'm thinking if there are too many fallbacks in this middleware. I think I'm fine with the first two but I'd remove this one since this is request *data* identification MW, so it shouldn't care about the request *host*.

Reply:
Isn't it two separate identification strategies?
1) by request header, 2) by query path?

Reply:
It is. But I'd say that having two MW classes for that would be unnecessary.

My thinking is that "You'd either use the header, or you'd not. In which case you would use the request data."

Hmm, I wonder what's the use case for even that fallback. I think everyone can make their HTTP client set request headers.

@JapSeyz Are you using the fallback to request data in your app?

I'd say that we can remove all the fallbacks and just use the header. If that doesn't work for someone, they can copy the class, change it to use request data and update their config file.

Unexpected fallbacks, similarly to unexpected global state, can result in confusing debugging.